### PR TITLE
Add misc macro for vector extensions.

### PR DESCRIPTION
**Pull request by kito-cheng**
Changes:
2021/08/19: Apply Fraser's fix.
2021/08/19: Change prefix to `__riscv_v_`

---

Background

RVV 1.0 has add few more extension (`zvl*b` and `zve*`) to give the promise about the EEW and VLEN, and this PR intend to expose those info by predefined marco.

https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#sec-vector-extensions

---
Introduce `__riscv_v_min_vlen`, `__riscv_v_max_eew` and `__riscv_v_max_eew_fp`
macro, for let programer easier to get vector related information, like:

- What's the minimal VLEN value is guaranteed current architecture
  extension?
- What's the maximal available EEW is guaranteed current
  architecture extension?
- What's the maximal available EEW for floating-point operation is
  guaranteed current architecture extension?

It's possible to get the information by testing serveral architecture extension
test macro, but the life can be simpler if toolchain can help, for
example, if you want to check the minimal VLEN value via those macro:

```c
 #if defined(__riscv_zvl512b)
   #define __riscv_v_min_vlen 512
 #elif defined(__riscv_zvl256b)
   #define __riscv_v_min_vlen 256
 #elif defined(__riscv_zvl128b) || defined(__riscv_v)
   #define __riscv_v_min_vlen 128
 #elif defined(__riscv_zvl64b) || defined(__riscv_zve64x) || \
       defined(__riscv_zve64f) || defined(__riscv_zve64d)
   #define __riscv_v_min_vlen 64
 #elif defined(__riscv_zvl32b) || defined(__riscv_zve32x) || \
       defined(__riscv_zve32f)
   #define __riscv_v_min_vlen 32
 #endif

```

And it's not scalable solution since in theory we can have up to zvl32768b...

### CODE REVIEW

**Comment by kito-cheng** [comment]
@rdolbeau @nick-knight @HanKuanChen @topperc @Hsiangkai @zakk0610 @JerryShih @rjiejie @rofirrim you might interested on this.

@jrtc27 @asb @luismarques you might not so interested, but we need LLVM folks :p

**Comment by HanKuanChen** [comment]
Because these macros belong to v extension, should not we add `_v` to the macros?
e.g., `__riscv_v_min_vlen`, `__riscv_v_max_eew` and `__riscv_v_max_eew_fp`

**Comment by rdolbeau** [comment]
Will there be some mechanism so that when compiling for zvl512b, the binary won't be able to run on a zvl256b-or-smaller machine?

**Comment by kito-cheng** [comment]
> Will there be some mechanism so that when compiling for zvl512b, the binary won't be able to run on a zvl256b-or-smaller machine?

We will encode `_zvl*b` in Tag_RISCV_arch@RISC-V attribute, and every executable and shared library will contain that info, so in theory we could use that to check on linux kernel and glibc, unfortunately both are not implement yet...but I believe we will have that in future.

[1] https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#list-of-attributes

**Comment by kito-cheng** [comment]
> Because these macros belong to v extension, should not we add _v to the macros?
> e.g., __riscv_v_min_vlen, __riscv_v_max_eew and __riscv_v_max_eew_fp

Sounds good to me, let me update :)

**Comment by kito-cheng** [comment]
Changes:
- Change prefix to `__riscv_v_`

File: riscv-c-api.md
Comment (frasercrmck) [inline riscv-c-api.md:51]:
```suggestion
extension requires VLEN ≥ 32 and `Zve64*` extension requires VLEN ≥ 64; also
```

File: riscv-c-api.md
Comment (frasercrmck) [inline riscv-c-api.md:100]:
```suggestion
- `__riscv_v_max_eew_fp` is 32 if `Zve32f` or `Zve64f` extension is available.
```
I don't think `Zve32d` exists, and even then it implies that FP64 is supported?

Comment (frasercrmck) [reply riscv-c-api.md:100]: I reversed to go from 32 to 64 but I just noticed the line below is still 64 to 32. Whatever's consistent.

Comment (kito-cheng) [reply riscv-c-api.md]: Thanks, I guess it might some thing wrong when I copy-past and modify :P

**Comment by kito-cheng** [comment]
Changes:
- Apply Fraser's suggestion.

File: riscv-c-api.md
Comment (Hsiangkai) [inline riscv-c-api.md:101]: Why does it define as 0, instead of undefined?

Comment (kito-cheng) [reply riscv-c-api.md:101]: My intention is make sure this marco is always defined if vector extension is available, so we only need to check `__riscv_v_max_eew_fp` is larger than our requirement once we ensure the vector extension is available.

e.g.
```
#ifdef __riscv_v_max_eew
#error "NO vector extension supported"
#endif

#if __riscv_v_max_eew_fp > 64
...
#endif
```

Rather than
```
#ifdef __riscv_v_max_eew
#error "NO vector extension supported"
#endif

#if defined(__riscv_v_max_eew_fp) && __riscv_v_max_eew_fp >= 64
...
#endif
```

**Comment by rjiejie** [comment]
+1

File: riscv-c-api.md
Comment (eopXD) [inline riscv-c-api.md:83]: I have a question on why `__riscv_v_max_eew` is undefined here. Although it seems that no vector functions is enabled if this macro is undefined. Maybe still set it as zero for consistency with `__riscv_v_max_eew_fp`?

Comment (kito-cheng) [reply riscv-c-api.md:83]: - `__riscv_v_max_eew_fp` is undefined if all of above rule are not match.
`__riscv_v_max_eew_fp` has same rule like this, undefined if no `V` or `Zve*` enabled.

Comment (eopXD) [reply riscv-c-api.md:83]: I see. Thank you for explaining.

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:73]: "macro"

**Comment by kito-cheng** [comment]
Changes:
- Fix typos.

**Comment by kito-cheng** [comment]
@aswaterman @kasanovic @nick-knight do you think `max_eew`/`max_eew_fp` is good enough or it should be `elen`/`elen_fp`?

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:62]:
```suggestion
The value of `__riscv_v_min_vlen` is defined by the following rules:

- 128, if the `V` extension is present;
- 32, if one of the `Zve32{x,f}` extensions is present;
- 64, if one of the `Zve64{x,f,d}` extensions is present;
- `N`, if one of the `Zvl<N>b` extensions, `N` in `{32,64,128,256,512,1024}`, is present.

If multiple rules apply, the maximum value is taken.
If none of the rules apply, `__riscv_v_min_vlen` is undefined.
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:64]:
```suggestion
Examples:
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:67]:
```suggestion
- `__riscv_v_min_vlen` is 512 for `rv32gcv_zvl512b`
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:68]:
```suggestion
- `__riscv_v_min_vlen` is 256 for `rv32gcv_zvl32b_zvl256b`
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:76]:
```suggestion
The `__riscv_v_max_eew` macro expands to the maximum EEW (effective element width), in bits, 
of any non-floating-point vector operand of any vector instruction in the available vector extension, if any.
(Stricter upper bounds may apply to particular operands of particular instructions.)
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:83]:
```suggestion
The value of `__riscv_v_max_eew` is defined by the following rules:

- 64, if the `V` extension or one of the `Zve64{x,f,d}` extensions is present; and
- 32, if one of the `Zve32{x,f}` extensions is present.

If multiple rules apply, the maximum value is taken.
If none of the rules apply, `__riscv_v_max_eew` is undefined.
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:86]: The technical term "vector type" is not currently defined in this document (or the C language) so I think we should avoid using/discussing it.

```suggestion
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:95]:
```suggestion
The `__riscv_v_max_eew_fp` macro expands to the maximum EEW (effective element width), in bits, 
of any floating-point vector operand of any vector instruction in the available vector extension, if any.
(Stricter upper bounds may apply to particular operands of particular instructions.)
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:102]:
```suggestion
The value of `__riscv_v_max_eew_fp` is defined by the following rules:

- 64, if one of the `V` or `Zve64d` extensions is present;
- 32, if one of the `Zve{32,64}f` extensions is present; and
- 0, if one of the `Zve{32,64}x` extensions is present.

If multiple rules apply, the maximum value is taken.
If none of the rules apply, `__riscv_v_max_eew_fp` is undefined.
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:106]: As above, the technical term "vector type" is not currently defined in this document (or the C language) so I think we should avoid using/discussing it.

```suggestion
```

File: riscv-c-api.md
Comment (nick-knight) [inline riscv-c-api.md:44]:
```suggestion
| __riscv_v_min_vlen    | <N> (see [__riscv_v_min_vlen](#__riscv_v_min_vlen)) | The `V` extension or one of the `Zve*` extension is available. |
| __riscv_v_max_eew     | <N> (see [__riscv_v_max_eew](#__riscv_v_max_eew)) | The `V` extension or one of the `Zve*` extensions is available. |
| __riscv_v_max_eew_fp  | <N> (see [__riscv_v_max_eew_fp](#__riscv_v_max_eew_fp)) | The `V` extension or one of the `Zve*` extensions is available. |
```

**Review by nick-knight — CHANGES_REQUESTED** [review]
@kito-cheng I have some suggestions for the wording and formatting. I don't think I've changed any of the semantics.

I still think we should have `__riscv_zve*` macros, but that can be a separate PR.

File: riscv-c-api.md
Comment (nick-knight) [reply riscv-c-api.md]: @kito-cheng My suggestion got broken somehow, I guess I must have been reviewing an older commit?

**Comment by kito-cheng** [comment]
@nick-knight 

> @kito-cheng My suggestion got broken somehow, I guess I must have been reviewing an older commit?

I got mail but not saw the message on github, anyway I'll merge those stuffs, thanks for your review comment!

**Comment by kito-cheng** [comment]
Changes:
- Apply suggestions from @nick-knight's code review

**Comment by topperc** [comment]
Did we reach consensus on whether `max_eew` or `elen` was the right term?

The spec defines ELEN as `The maximum size in bits of a vector element that any operation can produce or consume`

Section 3.4.2 includes these sentences `In the standard extensions, SEWMIN=8. For standard vector extensions with
ELEN=32, fractional LMULs of 1/2 and 1/4 must be supported. For standard vector extensions with ELEN=64, fractional
LMULs of 1/2, 1/4, and 1/8 must be supported.`

I take "standard vector extensions with ELEN=32" to mean Zve32x, Zve32f.

And "standard vector extensions with ELEN=64" to mean Zve64x, Zve64f, Zve64d, and V.

So to me it feels like `elen` is the right term for the concept.

**Comment by kasanovic** [comment]
ELEN works for the purpose of knowing the largest size individual element.   Some instructions, like SiFive matrix ops, effectively operate on larger units (e.g., 256b) but they are treated as vectors/aggregates of smaller elements, with the aggregate delimited by using larger vl rather than a larger SEW/EEW.  There will be some debate as we add vector crypto on how to handle this for long crypto blocks, and not sure how that will land yet.

**Comment by kito-cheng** [comment]
Sounds like ELEN is more right term here, let me update that.

**Comment by kito-cheng** [comment]
Changes:
- Rename `__riscv_v_max_eew`/`__riscv_v_max_eew_fp` to `__riscv_v_elen`/`__riscv_v_elen_fp`

**Comment by kito-cheng** [comment]
@aswaterman @kasanovic @nick-knight  let me know if look good to you :)

**Comment by nick-knight** [comment]
@kito-cheng I don't see any harm in the current proposal, as long as we still plan to add `zve*` macros (perhaps in a separate PR).

**Comment by kito-cheng** [comment]
@nick-knight Sure, macro for `zve*` extension will added in separated PR.

**Review by nick-knight — APPROVED** [review]

**Comment by kito-cheng** [comment]
@asb @frasercrmck @topperc @rdolbeau would you mind give an explicit LGTM if that's OK to you? thanks :)

**Comment by frasercrmck** [comment]
Looks good other than minor clarifications

File: riscv-c-api.md
Comment (topperc) [inline riscv-c-api.md:44]: The when defined for `__riscv_v_elen_fp` isn't quite right. zve32x/zve64x don't define this.

Comment (kito-cheng) [reply riscv-c-api.md:44]: `zve32x` and `zve64x` still define that but defined as `0`

Comment (topperc) [reply riscv-c-api.md:44]: Oops I missed that. Thanks.

**Comment by asb** [comment]
> @asb @frasercrmck @topperc @rdolbeau would you mind give an explicit LGTM if that's OK to you? thanks :)

Craig and Fraser have done much more with the vector spec than I have, so I'm happy if they're happy.

**Comment by topperc** [comment]
LGTM

**Comment by kito-cheng** [comment]
Gonna merge this, we got approval from GNU and LLVM community and vector ext. are ratified ext.

**Pull request merged**
